torch/csrc/jit/codegen/onednn/graph_helper.cpp

```diff
@@ -0,0 +1,561 @@
+          .setAttr("pads_end", Operator::Ints, 3)
+          .setAttr("exclude_pad", !Operator::Bool(node, 5))
+          .setAttr("rounding_type", std::string(rounding_type));
+    }
+
+    case aten::matmul: {
+      auto dim0 = getDimensions(node->namedInput("self")).value_or(-1);
+      auto dim1 = getDimensions(node->namedInput("other")).value_or(-1);
+      // TODO: support all shape combinations
+      REQUIRE(
+          (dim0 == 2 && dim1 == 2) || (dim0 == 4 && dim1 == 4) ||
+          (dim0 == 3 && dim1 == 2));
+    } // fall through
+    case aten::mm: {
+      return Operator(node, opkind::MatMul).setInput(0, 1).setOutput(0);
+    }
+
+    case aten::linear: {
+      return Operator(node, opkind::MatMul)
+          .setInput(0, 1, 2)
+          .setOutput(0)
+          .setAttr("transpose_b", true);
+    }
+
+    default:
+      return makeWildcardOp(node);
+  }
+}
+
+dnnl::graph::op createLlgaOp(Node* node) {
+  return createOperator(node).llgaOp();
+}
+
+bool isSupported(Node* node) {
+  return createOperator(node).kind() != opkind::Wildcard;
+};
+
+DeviceType inferDeviceFromValue(Value* v) {
+  auto tt = v->type()->cast<TensorType>();
+  if (!tt) {
+    return at::kCPU;
+  }
+  auto device = tt->device();
+  if (!device) {
+    return at::kCPU;
+  }
+  return device->type();
+}
+
+DeviceType inferDevice(const std::shared_ptr<Graph>& graph) {
+  auto dt = inferDeviceFromValue(graph->inputs()[0]);
+  TORCH_CHECK(
+      std::all_of(
+          graph->inputs().begin(),
+          graph->inputs().end(),
+          [dt](Value* v) { return inferDeviceFromValue(v) == dt; }),
+      "All inputs must have the same deive type");
+  return dt;
+}
+
+dnnl::graph::engine::kind getLlgaEngineKind(DeviceType type) {
+  switch (type) {
+    case DeviceType::CPU:
+      return dnnl::graph::engine::kind::cpu;
+    default:
+      TORCH_CHECK(false, "Not support device type ", type);
+  }
+}
+
+void mayAddListConstructIntoConcatPartition(
+    Node* n,
+    OpPartitionMap& opToOwningPartition) {
+  // Since prim::ListConstruct is not visible to the LLGA,
+  // it will not be in any partition returned from partfuseritioning results.
+  // We need rewrite opToOwningPartition to make the prim::ListConstruct to be
+  // 'virtually' in the same partition with the aten::cat, so that
+  // prim::ListConstruct can be fused into the fusion group by graph fuser.
+  // We emphasize on 'virtually' because get_num_ops() for cat's partition
+  // would still return 1.
+  if (n->kind() == aten::cat && opToOwningPartition.has(n)) {
+    auto listConstrcut = n->namedInput("tensors")->node();
+    auto partitionId = opToOwningPartition.get(n);
+    opToOwningPartition.add(listConstrcut, partitionId);
+  }
+}
+
+// Verify that input tensors are compatible with oneDNN Graph.
+// Scalars would be converted to 1-D tensors later anyway,
+// but they shouldn't be complex-double
+// If this check fails, convert op to wildcard
```

Comment:
nit: still don't like the fact that we know that this will fail and then still try to send it to OneDNN, it makes debugging why something wasn't failed harder to follow